ethers-core/src/macros/ethers_crate.rs

```diff
@@ -1,123 +1,310 @@
+        }
+    }
+
+    /// `<ethers_*>`
+    #[inline]
+    pub fn path_names() -> impl Iterator<Item = (Self, &'static str)> {
+        Self::iter().map(|x| (x, x.path_name()))
     }
 
-    res
+    /// `<ethers::*>`
+    #[inline]
+    pub fn ethers_path_names() -> impl Iterator<Item = (Self, &'static str)> {
+        Self::iter().map(|x| (x, x.ethers_path_name()))
+    }
+
+    /// Returns the [`Path`][syn::Path] in the current project.
+    #[inline]
+    pub fn get_path(&self) -> syn::Path {
+        let name = ETHERS_CRATE_NAMES[self];
+        syn::parse_str(name).unwrap()
+    }
+}
+
+/// Returns whether `crate`, in the current environment, refers to the root package.
+///
+/// This is false for integration tests, benches, and examples, as the `crate` keyword will not
+/// refer to the root package.
+///
+/// We can find this using some [environment variables set by Cargo during compilation][ref]:
+/// - `CARGO_TARGET_TMPDIR` is only set when building integration test or benchmark code;
+/// - When `CARGO_MANIFEST_DIR` contains `/benches/` or `/examples/`
+/// - `CARGO_CRATE_NAME`, see [`is_crate_name_in_dirs`].
+///
+/// [ref]: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates
+#[inline]
+fn is_crate_root(manifest_dir: impl AsRef<Path>) -> bool {
+    let manifest_dir = manifest_dir.as_ref();
+    env::var_os("CARGO_TARGET_TMPDIR").is_none() &&
+        manifest_dir.components().all(|c| {
+            let s = c.as_os_str();
+            s != "examples" && s != "benches"
+        }) &&
+        !is_crate_name_in_dirs(manifest_dir)
+}
+
+/// Returns whether `CARGO_CRATE_NAME` is the name of a file or directory in the first level of
+/// `manifest_dir/{benches,examples,tests}/`.
+///
+/// # Example
+///
+/// With this project structure:
+///
+/// ```text
+/// .
+/// ├── Cargo.lock
+/// ├── Cargo.toml
+/// ├── src/
+/// │   ...
+/// ├── benches/
+/// │   ├── large-input.rs
+/// │   └── multi-file-bench/
+/// │       ├── main.rs
+/// │       └── bench_module.rs
+/// ├── examples/
+/// │   ├── simple.rs
+/// │   └── multi-file-example/
+/// │       ├── main.rs
+/// │       └── ex_module.rs
+/// └── tests/
+///     ├── some-integration-tests.rs
+///     └── multi-file-test/
+///         ├── main.rs
+///         └── test_module.rs
+/// ```
+///
+/// The resulting `CARGO_CRATE_NAME` values will be:
+///
+/// |                  Path                  |          Value         |
+/// |:-------------------------------------- | ----------------------:|
+/// | benches/large-input.rs                 |            large-input |
+/// | benches/multi-file-bench/\*\*/\*.rs    |       multi-file-bench |
+/// | examples/simple.rs                     |                 simple |
+/// | examples/multi-file-example/\*\*/\*.rs |     multi-file-example |
+/// | tests/some-integration-tests.rs        | some-integration-tests |
+/// | tests/multi-file-test/\*\*/\*.rs       |        multi-file-test |
```

Comment:
This is not entirely true, but I couldn't really find a good way to get these values..

Reply:
fine w/ me